### PR TITLE
Fix incorrect error reporting from Fast Sort check

### LINKED ISSUE #34: Exception is caught always for sort option
Repro: 
1. Disable/exit Everything
2. Use Everything in flow
3. The error message is fast sort option not enabled instead of service is not running.

### PR DESCRIPTION
**Pull request by jjw24**
- Fix incorrect error reporting from Fast Sort check, which reports service not running error as fast sort option not enabled.
- Updated Fast Sort option not supported message
- Switch the non-Fast Sort search restriction to plugin setting warning
- Show Everything service not running in settings
image
image

close https://github.com/Flow-Launcher/Flow.Launcher.Plugin.Everything/issues/34

### CODE REVIEW

File: Everything/EverythingAPI.cs [hunk 1]
```diff
@@ -123,11 +123,13 @@ public List<SearchResult> Search(string keyWord, CancellationToken token, SortOp
                 EverythingApiDllImport.Everything_SetSearchW(keyWord);
                 EverythingApiDllImport.Everything_SetOffset(offset);
                 EverythingApiDllImport.Everything_SetMax(maxCount);
-                
-                if(!EverythingApiDllImport.Everything_IsFastSort(sortOption))
-                {
+
+                var fastSortOptionEnabled = EverythingApiDllImport.Everything_IsFastSort(sortOption);
+                CheckAndThrowExceptionOnError();
```
Comment (taooceros) [inline Everything/EverythingAPI.cs]: What does this method do?

**Comment by jjw24** [comment]
@taooceros I have not been successful in disabling fast sort option to check this condition:

image

Turning off all these still get the Fast Sort option as true . Any ideas?

**Comment by taooceros** [comment]
I do notice this error for quite a while, but forget to fix it before😯

**Comment by taooceros** [comment]
Name is fast sort by default.

File: Everything/EverythingAPI.cs (hunk 1 above)
Comment (jjw24) [reply Everything/EverythingAPI.cs]: it checks the error thrown by Everything via API, it's how the code checks for service not running. The service not running error is 'silent'.

https://www.voidtools.com/support/everything/sdk/everything_isfastsort/, see EVERYTHING_ERROR_IPC

**Comment by taooceros** [comment]
Probably you may need a rebuild index before seeing the effect of fast sort (because it is basically an index i guess)

**Comment by jjw24** [comment]
> Name is fast sort by default.

hmm which one is not do you know?
image

**Comment by taooceros** [comment]
Hmm I think if there's an option, it shouldn't be fast sort by default.

**Comment by jjw24** [comment]
> Hmm I think if there's an option, it shouldn't be fast sort by default.

sorry what do you mean?

**Comment by taooceros** [comment]
And there are a few that seems cannot be fast sort (filelist I guess).

**Comment by taooceros** [comment]
> > Hmm I think if there's an option, it shouldn't be fast sort by default.

Oh as I remember, only name is fast sort by default. 
I mean the option to enable fast sort in the previous image.

**Comment by jjw24** [comment]
I have switched them off and still not getting the error lol

image

**Comment by taooceros** [comment]
Try force rebuild?

**Comment by taooceros** [comment]
Have you changed the sort option in Flow? The default sort option in Flow is name, which is always fast sort.

**Comment by jjw24** [comment]
> Have you changed the sort option in Flow? The default sort option in Flow is name, which is always fast sort.

that's it lol, i did not notice the option in flow, rather was trying with Everything's own settings, and yep File List FileName sort option is not fast sort.

I will change the error message.

**Comment by jjw24** [comment]
should we remove File List FileName sort option from being set in flow's settings if we dont allow it?

**Comment by jjw24** [comment]
For prosperity, these are the non-Fast Sort options in Everything:

- File List FileName sort option
- Path
- Size
- Type Name

**Comment by jjw24** [comment]
@taooceros i tried a couple of the non-Fast Sort options, seems to work ok, obviously depends on the number of results the user gets, but do we want to allow those options and just put the warning in the settings page next to the sort drop down options?

**Comment by taooceros** [comment]
Sorry I didn't respond early because I miss the message.

> @taooceros i tried a couple of the non-Fast Sort options, seems to work ok, obviously depends on the number of results the user gets, but do we want to allow those options and just put the warning in the settings page next to the sort drop down options?

Yeah we can, but Flow seems currently don't support FileList, so maybe remove that for now?

**Comment by taooceros** [comment]
@jjw24 Size is one of the fast sort option, same as Path. FileList FileName may not exist without extra setting, so I think we can remove it for now unless user want it.

From my testing, Type Name is not much slower than other FastSort (probably extension is indexed, so even its fastsort option is not enabled, it is not slow).

**Comment by jjw24** [comment]
So sounds like if we remove File list Filename sort then we can remove that restriction then? No need for any warning

**Comment by taooceros** [comment]
> So sounds like if we remove File list Filename sort then we can remove that restriction then? No need for any warning

Hm not sure whether I understand here idea clearly. I add restrictions because everything include every files in the disk, so enable sorting without index will take almost forever. It is not quite possible to use non fastsort in Flow. Most fastsort is not enabled by default in everything, so a warning for user is required I think. 
The Filelist Filename has nothing to do with this, but just I am not quite sure how to use it though. 😂

**Comment by jjw24** [comment]
Oh ok, so we remove FileList Filename and add the warning to plugin's settings and allow them to use any options then?

Basically that method that checks if the sort option is fast sort, I will move it out of the main search method and just display a warning base on the fast sort check result

**Comment by taooceros** [comment]
Have you checked what will happen without fast search? In my computer, using not fast search will cause a huge cpu usage and computer slow down, and results will not be returned unless wait for a long time.

**Comment by jjw24** [comment]
So I tried a couple of them before (earlier comment), seems fine on my computer. Which non-fast sort option is causing the slow down for u can u list them please.

**Comment by taooceros** [comment]
Would you please take a try on attributes and deselect the option of indexing attribute, and force rebuild?

**Comment by jjw24** [comment]
so i tested those non-fast sort options and the result for me as i dont have a numerous amount of files to search through, using these options are fine. I did these steps:

**commented out the code that throws the warning**

**1. deselect all these fast options**
image

**2. change sort to the non-fast option**
image

**3. force index rebuild**
image

**4. Change the settings in Flow's Everything plugin to match**
image
 
**5 repeat above for the other options**

**Comment by jjw24** [comment]
since it seems ok with a smaller number of results, do we want to move the warning to the Plugin's settings Sort By option so users can still use them, or remove those non-fast sort option from being able to be selected in flow so we dont need to display the warning message. What do you think @taooceros ?

**Comment by taooceros** [comment]
Yeah if that works, we can do that. Probably because my computer got too much files so it stop responding for a long time when non fast sort.

File: EverythingSettings.xaml [hunk 2]
```diff
@@ -66,5 +68,10 @@
                 </DataTemplate>
             </ComboBox.ItemTemplate>
         </ComboBox>
+
+        <TextBlock Name ="tbFastSortWarning"  Grid.Row="4" Grid.Column="2" Margin="10 5 10 0" TextAlignment="Left"
+                   Text="{DynamicResource flowlauncher_plugin_everything_nonfastsort_warning}"
```
Comment (taooceros) [inline EverythingSettings.xaml]: We may also need to notify user when everything is not running (which we cannot check whether the option is fast sort except sorting by name) .

Comment (jjw24) [reply EverythingSettings.xaml]: done

**Review by taooceros — APPROVED** [review]

**Pull request merged**